### PR TITLE
Add more detailed image validator

### PR DESCRIPTION
**Pull request by adamjtaylor**
This PR adds validation method `tiffinffo` which performes more detailed validation of images, particularly for risk of PHI

- Warning if fails `xmlvalid`
- Error if string `date` detected in `tiffinfo` output
- Warning string `time` detected in `tiffinfo` output
- Warning label image detected in `tiffinfo` output

### CODE REVIEW

File: htan_registry/tiffinfo.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+        # Validate the OME-XML with xmlvalid
+        xmlvalid_cmd = os.path.join(here, '../bftools/xmlvalid')
+        xmlvalid_exc = subprocess.run([xmlvalid_cmd, path_or_df], capture_output=True, text=True)
+        # If error, set the set a warning
+        if xmlvalid_exc.returncode:
+            warning.append("Invalid OME-XML")
+
+        # Get all information on the file from tiffdump
+        tiffdump_exc = subprocess.run(["tiffinfo", path_or_df], capture_output=True, text=True)
```
Comment (thomasyu888) [inline htan_registry/tiffinfo.py]: What is `tiffinfo`?  Is it part of `bftools`?

File: htan_registry/tiffinfo.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+                                     toDelete=True)
+        df.to_csv(newPath, sep="\t", index=False)
+        return newPath
+
+    def _validate(self, path_or_df):
+        total_error = []
+        warning = []
+
+        # Validate the OME-XML with xmlvalid
+        xmlvalid_cmd = os.path.join(here, '../bftools/xmlvalid')
```
Comment (thomasyu888) [inline htan_registry/tiffinfo.py]: Whats the difference between this and the ometif class?

File: htan_registry/tiffinfo.py (hunk 1 above)
Comment (adamjtaylor) [reply htan_registry/tiffinfo.py]: It is part of `libtiff-tools` and can be used to print out information from the headers
> Tiffinfo displays information about files created according to the Tag Image File Format, Revision 6.0. By default, the contents of each TIFF directory in each file is displayed, with the value of each tag shown symbolically (where sensible).

Install with `sudo apt install libtiff-tools` or equivalent

Comment (adamjtaylor) [reply htan_registry/tiffinfo.py]: We can add this to requirements or documentation

File: htan_registry/tiffinfo.py (hunk 2 above)
Comment (adamjtaylor) [reply htan_registry/tiffinfo.py]: `ometif` class to explicitly validate `ome-tiff` files. `tiffinfo` class intended to validate and basic PHI check any tiff based file. Can have as one class if better.

Comment (thomasyu888) [reply htan_registry/tiffinfo.py]: Hmmm. Let's discuss more in our chat.  Basically what we can have are more modular validation functions in a separate module and import them.  Then you have different file format classes like `class Svs()` and `class OmeTiff()` and `class Tiff()` that can call the same functions, BUT if you wanted to add extra validation specific to each of those file formats, you then could.

File: htan_registry/tiffinfo.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        tiffinfo_exc = subprocess.run(["tiffinfo", path_or_df], capture_output=True, text=True)
+        # If error, set the error
+        if tiffinfo_exc.returncode:
+            total_error.append("tiffinfo failed: {tiffinfo_exc.stdout}")
+        # If dates found, make an error
+        finddate = len(re.findall(r'date', tiffinfo_exc.stdout, re.IGNORECASE))
+        # If likely redacted dates are found make a warning instead
+        redacted_dates = len(re.findall(r'1970:01:01 00:00:00', tiffinfo_exc.stdout))
+        if redacted_dates:
+            warning.append(f'{finddate} occurences of DATE in TIFF headers that are potentially redacted to "1970:01:01 00:00:00"')
```
Comment (thomasyu888) [inline htan_registry/tiffinfo.py]: Warnings currently aren't as well returned.  I don't think I return an email for valid uploads just yet, so an email doesn't get sent out about errors.  I'm not 100% certain, but warnings also currently aren't stored if the file is valid.

File: htan_registry/tiffinfo.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+from synapsegenie.example_filetype_format import FileTypeFormat
+from synapsegenie import process_functions
+
+logger = logging.getLogger(__name__)
+
+here = os.path.abspath(os.path.dirname(__file__))
+
+
+class TiffInfo(FileTypeFormat):
```
Comment (thomasyu888) [inline htan_registry/tiffinfo.py]: Make this imaging class and remove OmeTif class

File: htan_registry/tiffinfo.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+    _filetype = "tiffinfo"
+
+    _process_kwargs = ["databaseSynId"]
+
+    def _validate_filetype(self, filePath):
+        assert os.path.basename(filePath[0]).endswith(".tif{1,2}|.svs")
```
Comment (thomasyu888) [inline htan_registry/tiffinfo.py]: ADd something like this
```
supported_file_ext = (".tif", ".tiff", ".svs", ".ndpi")
assert os.path.basename(filePath[0]).endswith(supported_file_ext)
```

File: htan_registry/tiffinfo.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+    def process_steps(self, df, newPath, databaseSynId):
+        df = self._process(df)
+        process_functions.updateData(self.syn, databaseSynId, df, self.center,
+                                     toDelete=True)
+        df.to_csv(newPath, sep="\t", index=False)
+        return newPath
+
+    def _validate(self, path_or_df):
+        total_error = []
```
Comment (thomasyu888) [inline htan_registry/tiffinfo.py]: Use `io.StringIO()`

File: htan_registry/tiffinfo.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+        # If error, set the set a warning
+        if showinf_exc.returncode:
+            total_error.append(f"showinf failed: {showinf_exc.stdout}")
+
+        # Validate the OME-XML with xmlvalid
+        xmlvalid_cmd = os.path.join(here, '../bftools/xmlvalid')
+        xmlvalid_exc = subprocess.run([xmlvalid_cmd, path_or_df], capture_output=True, text=True)
+        # If error, set the set a warning
+        if xmlvalid_exc.returncode:
+            warning.append("xmlvalid failed: {xmlvalid_exc.stdout}")
```
Comment (thomasyu888) [inline htan_registry/tiffinfo.py]: Add `\n`

Comment (thomasyu888) [inline htan_registry/tiffinfo.py]:
```suggestion
            warning.append(f"xmlvalid failed: {xmlvalid_exc.stdout}\n")
```

**Review by thomasyu888 — APPROVED** [review]

**Pull request merged**